app/scripts/modules/core/src/overrideRegistry/Overridable.tsx

```diff
@@ -0,0 +1,144 @@
+ * If yes, the overriding component is rendered.
+ * If no, the decorated component itself is rendered.
+ *
+ * @Overridable('overrideKey')
+ * class MyCmp extends React.Component {
+ *   render() { return <h1>Overridable Component</h1> }
+ * }
+ *
+ * When using the component, just render it as usual:
+ * <MyCmp/>
+ *
+ * If the override is cloud provider specific, pass the accountId as a prop:
+ * <MyCmp accountId={accountId} />
+ */
+export function Overridable(key: string) {
+  return function <P extends IOverridableProps, T extends React.ComponentClass<P>> (targetComponent: T): T {
+    return overridableComponent(targetComponent, key)
+  }
+}
+
+/**
+ * A high order component which returns a delegating component.
+ * The component will delegate to the overriding component registered with OverrideRegistry or CloudProviderRegistry.
+ * If no override is registered, it delegates to the component being decorated.
+ *
+ * class MyCmp extends React.Component {
+ *   render() { return <h1>Overridable Component</h1>
+ * }
+ *
+ * export const MyOverridableCmp = overridableComponent(MyCmp);
+ */
+export function overridableComponent <P extends IOverridableProps, T extends React.ComponentClass<P>> (DefaultComponent: T, key: string): T {
+  class OverridableComponent extends React.Component<P, { Component: T }> {
+    private account$ = new Subject<string>();
+    private destroy$ = new Subject();
+
+    constructor(props: P) {
+      super(props);
+
+      const { accountService } = ReactInjector;
+
+      this.account$
+        .switchMap(accountId => {
+          if (accountId) {
+            return Observable.fromPromise(accountService.getAccountDetails(accountId));
+          }
+
+          return Observable.of(null);
+        })
+        .map((accountDetails: IAccountDetails) => this.getComponent(accountDetails))
+        .takeUntil(this.destroy$)
+        .subscribe((Component) => this.setState({ Component }));
+    }
+
+    public componentDidMount() {
+      this.account$.next(this.props.accountId);
+    }
+
+    public componentWillUnmount() {
+      this.destroy$.next();
+    }
+
+    public componentWillReceiveProps(nextProps: P) {
+      const { accountId } = nextProps;
+      if (this.props.accountId !== accountId) {
+        this.account$.next(accountId);
+      }
+    }
+
+    private getComponentFromCloudProvider(accountDetails: IAccountDetails): T {
+      const { cloudProvider, providerVersion } = accountDetails;
+      const { cloudProviderRegistry } = ReactInjector;
+      if (!cloudProvider) {
+        return null;
+      }
+
+      const CloudProviderComponentOverride = cloudProviderRegistry.getValue(cloudProvider, key, providerVersion);
+      if (CloudProviderComponentOverride) {
+        return CloudProviderComponentOverride as T;
+      }
+
+      const cloudProviderTemplateOverride = cloudProviderRegistry.getValue(cloudProvider, key + 'TemplateUrl', providerVersion);
```

Comment:
Check for templateUrl + controller second.  These follow a naming convention `key + 'Controller'` or `key + 'TemplateUrl'`